### PR TITLE
feat: specify the resource requirements for the kaniko pod

### PR DESCRIPTION
**Pull request by garethjevans**

### CODE REVIEW

**Comment by garethjevans** [comment]
This PR allows the user to specify the limits/requests values for the kaniko pod during a build.  Our testing has found this is a more effective way of allowing the cluster to autoscale correctly.

If you're happy with this PR, let me know and I'll get some docs added.

File: pkg/skaffold/schema/latest/config.go [hunk 1]
```diff
@@ -268,6 +268,9 @@ type KanikoBuild struct {
 	// DockerConfig describes how to mount the local Docker configuration into the
 	// Kaniko pod.
 	DockerConfig *DockerConfig `yaml:"dockerConfig,omitempty"`
+
+	// ResourceRequirements define the resource requirements for the kaniko pod
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Can you put a final dot here? I really need to add a unit test to validate those comments.

Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Comments should start by the name of the field.

File: pkg/skaffold/schema/latest/config.go [hunk 2]
```diff
@@ -279,6 +282,22 @@ type DockerConfig struct {
 	SecretName string `yaml:"secretName,omitempty"`
 }
 
+// ResourceRequirements describes the resource requirements for the kaniko pod
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Can you put a final dot?

File: pkg/skaffold/schema/latest/config.go [hunk 3]
```diff
@@ -279,6 +282,22 @@ type DockerConfig struct {
...
+type ResourceRequirements struct {
+	// Requests https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Can you add a short description and use Mardown notation for the link?

File: pkg/skaffold/schema/latest/config.go [hunk 4]
```diff
@@ -279,6 +282,22 @@ type DockerConfig struct {
...
+	Requests *ResourceRequirement `yaml:"requests,omitempty"`
+	// Limits https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Same here

File: pkg/skaffold/schema/latest/config.go [hunk 5]
```diff
@@ -279,6 +282,22 @@ type DockerConfig struct {
...
+	Limits *ResourceRequirement `yaml:"limits,omitempty"`
+}
+
+// ResourceRequirement see https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Also here

File: pkg/skaffold/schema/latest/config.go [hunk 6]
```diff
@@ -279,6 +282,22 @@ type DockerConfig struct {
...
+type ResourceRequirement struct {
+	CPU    string `yaml:"cpu,omitempty"`
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Can you add comments for CPU and Memory?

File: pkg/skaffold/schema/latest/config.go (hunk 1 above)
Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: done

Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: done

File: pkg/skaffold/schema/latest/config.go (hunk 2 above)
Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: done

File: pkg/skaffold/schema/latest/config.go (hunk 3 above)
Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: done

File: pkg/skaffold/schema/latest/config.go (hunk 5 above)
Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: done

File: pkg/skaffold/schema/latest/config.go (hunk 4 above)
Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: done

File: pkg/skaffold/schema/latest/config.go (hunk 6 above)
Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: done

File: pkg/skaffold/build/kaniko/sources/sources_test.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			actual := podTemplate(test.initial, test.image, test.args)
+			testutil.CheckDeepEqual(t, test.expected, actual, opt)
+		})
+	}
+}
+
+func createResourceRequirements(cpuLimit resource.Quantity, memoryLimit resource.Quantity, cpuRequest resource.Quantity, memoryRequest resource.Quantity) v1.ResourceRequirements {
```
Comment (dgageot) [inline pkg/skaffold/build/kaniko/sources/sources_test.go]: Should be something like:
```
func createResourceRequirements(cpuLimit resource.Quantity, memoryLimit resource.Quantity, cpuRequest resource.Quantity, memoryRequest resource.Quantity) v1.ResourceRequirements {
	return v1.ResourceRequirements{
		Limits: v1.ResourceList{
			v1.ResourceCPU:    cpuLimit,
			v1.ResourceMemory: memoryLimit,
		},
		Requests: v1.ResourceList{
			v1.ResourceCPU:    cpuRequest,
			v1.ResourceMemory: memoryRequest,
		},
	}
}
```

File: pkg/skaffold/schema/latest/config.go [hunk 8]
```diff
@@ -254,6 +254,9 @@ type ClusterDetails struct {
 
 	// DockerConfig describes how to mount the local Docker configuration into a pod.
 	DockerConfig *DockerConfig `yaml:"dockerConfig,omitempty"`
+
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: Those changes will require introducing a `v1beta8` schema

File: pkg/skaffold/schema/latest/config.go [hunk 9]
```diff
@@ -265,6 +268,26 @@ type DockerConfig struct {
 	SecretName string `yaml:"secretName,omitempty"`
 }
 
+// ResourceRequirements describes the resource requirements for the kaniko pod.
+type ResourceRequirements struct {
+	// Requests resource requests for the Kaniko pod. [More on requests](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container).
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: nit: The link could be on `resource requests`. That makes the online doc shorter and easier to read.

File: pkg/skaffold/schema/latest/config.go [hunk 10]
```diff
@@ -265,6 +268,26 @@ type DockerConfig struct {
...
+	Requests *ResourceRequirement `yaml:"requests,omitempty"`
+
+	// Limits resource limits for the Kaniko pod. [More on limits](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container).
```
Comment (dgageot) [inline pkg/skaffold/schema/latest/config.go]: nit: The link could be on `resource limits`. That makes the online doc shorter and easier to read.

File: pkg/skaffold/build/kaniko/sources/sources_test.go [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/testutil"
+)
+
+func TestPodTemplate(t *testing.T) {
+	tests := []struct {
+		name     string
```
Comment (dgageot) [inline pkg/skaffold/build/kaniko/sources/sources_test.go]: nit: This is usually called `description`

File: pkg/skaffold/build/kaniko/sources/sources_test.go (hunk 7 above)
Comment (garethjevans) [reply pkg/skaffold/build/kaniko/sources/sources_test.go]: done

File: pkg/skaffold/build/kaniko/sources/sources_test.go (hunk 11 above)
Comment (garethjevans) [reply pkg/skaffold/build/kaniko/sources/sources_test.go]: done

File: pkg/skaffold/schema/latest/config.go (hunk 9 above)
Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: Done

File: pkg/skaffold/schema/latest/config.go (hunk 10 above)
Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: Done

File: pkg/skaffold/schema/latest/config.go (hunk 8 above)
Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: @dgageot should the new schema version be introduced as part of this change or separately?

Comment (garethjevans) [reply pkg/skaffold/schema/latest/config.go]: actually, ignore that comment, I can see that a new schema version has already been introduced.

**Comment by tejal29** [comment]
>This PR allows the user to specify the limits/requests values for the kaniko pod during a build. Our testing has found this is a more effective way of allowing the cluster to autoscale correctly.

Thanks @garethjevans for this feature request. Since this a config change, we need to have a better understanding of this.
Can you shed some light on:-
1. What were the other alternatives other than using the config?
1. How much is the improvement?
1. What is the default limit/request value if not specified?

**Comment by garethjevans** [comment]
Hi @tejal29,

Some background: we are trying to use skaffold on jenkins-x to build many images concurrently, sometimes in excess of 50 concurrently.  With no resources limits set these often get scheduled on the same node, leading to them being evicted under node pressure.

1) What were the other alternatives other than using the config?

Without this change the only way we've managed to stop Kaniko pods getting Evicted under load is the create a resource quota (https://kubernetes.io/docs/concepts/policy/resource-quotas/) for the namespace.  Unfortunately, using this approach, it doesn't allow us to customise each pod.

2) How much is the improvement?

We've found during our testing that k8s is able to schedule containers far better when they have pod resource limts / requests correctly configured.

3) What is the default limit/request value if not specified?

By default, if not specified, we revert to the original behaviour of not specifying any resources limits. This may work fine if you building images frequently, but not when the nodes are under load.

**Comment by tejal29** [comment]
Thanks @garethjevans  this makes sense to me now. 
I am thinking this should be the case whenever skaffold create Pods? and maybe should be a top level config.  I did a quick grep on master and found there are few places
```tejaldesai@tejaldesai~/go/src/github.com/GoogleContainerTools/skaffold: (master)$ git grep 'v1.PodSpec' | uniq
pkg/skaffold/build/kaniko/sources/sources.go:        Spec: v1.PodSpec{
pkg/skaffold/debug/debug_test.go:        Spec:       v1.PodSpec{Containers: []v1.Container{{Name: "name1", Image: "image1"}}}}
pkg/skaffold/debug/transform.go:func transformPodSpec(metadata *metav1.ObjectMeta, podSpec *v1.PodSpec, retrieveImageConfiguration configurationRetriever) bool {
pkg/skaffold/debug/transform.go:func allocatePort(podSpec *v1.PodSpec, desiredPort int32) int32 {
pkg/skaffold/debug/transform.go:func isPortAvailable(podSpec *v1.PodSpec, port int32) bool {
pkg/skaffold/debug/transform_jvm_test.go:                Spec: v1.PodSpec{Containers: []v1.Container{
pkg/skaffold/debug/transform_jvm_test.go:                        Spec: v1.PodSpec{Containers: []v1.Container{
pkg/skaffold/debug/transform_nodejs_test.go:                Spec: v1.PodSpec{Containers: []v1.Container{
pkg/skaffold/debug/transform_nodejs_test.go:                        Spec: v1.PodSpec{Containers: []v1.Container{
pkg/skaffold/debug/transform_test.go:        pod         v1.PodSpec
pkg/skaffold/debug/transform_test.go:            pod:         v1.PodSpec{},
pkg/skaffold/debug/transform_test.go:            pod: v1.PodSpec{Containers: []v1.Container{
pkg/skaffold/debug/transform_test.go:            pod:         v1.PodSpec{},
pkg/skaffold/debug/transform_test.go:            pod: v1.PodSpec{Containers: []v1.Container{
pkg/skaffold/kubernetes/colorpicker_test.go:                Spec: v1.PodSpec{
pkg/skaffold/kubernetes/port_forward_test.go:                    Spec: v1.PodSpec{
pkg/skaffold/sync/sync_test.go:    Spec: v1.PodSpec{
pkg/webhook/kubernetes/deployment.go:                Spec: v1.PodSpec{
```
But digging more into deployment.go to see what it does.

**Comment by tejal29** [comment]
ah nvm, ignore the "pkg/webhook/kubernetes/deployment.go" it is something internal for generating docs.

**Comment by tejal29** [comment]
@garethjevans I discussed this with the team and this looks good!
Since this is a config change, we have to freeze the config.
I am going to get that Pr out and then once that is in, please rebase and this PR will be ready to be merged!!

**Comment by tejal29** [comment]
@garethjevans Please rebase and this should be ready to merge!

**Comment by garethjevans** [comment]
@tejal29 the PR has been rebased and is ready to go

**Pull request merged**